Kumi.Game/Gameplay/Mods/Mod.cs

```diff
@@ -0,0 +1,35 @@
+﻿using Kumi.Game.Screens.Play;
+using osu.Framework.Timing;
+
+namespace Kumi.Game.Gameplay.Mods;
+
+/// <summary>
+/// An abstract class that represents a modification to the gameplay.
+/// </summary>
+public abstract class Mod
+{
+    /// <summary>
+    /// The name of the mod.
+    /// </summary>
+    public abstract string Name { get; }
+    
+    /// <summary>
+    /// The named tag of the mod, usually an abbreviation of the name.
+    /// </summary>
+    public abstract string Tag { get; }
+
+    /// <summary>
+    /// A description of the mod, used to explain what the mod does.
+    /// </summary>
+    public virtual string Description => "No description provided.";
+    
+    /// <summary>
+    /// A list of mods that are incompatible with this mod.
+    /// </summary>
+    public virtual Type[] IncompatibleMods => new Type[0];
+    
+    /// <summary>
+    /// Applies the mod to the player.
+    /// </summary>
+    public abstract void Apply(Player player, IAdjustableClock clock);
```

Comment:
I'm not sure how I feel about having the core mod class having this. Other mod types might have to change the chart itself, for example, which we could pre-load in the `PlayerLoader` screen. Or for rate-adjust mods, which we might want to audibly show just how fast the song is going.

This could be done by using interfaces instead, then filtering out through each active mod and applying the function accordingly.

Reply:
Then add that to the function itself. I'm not going to inevitably add more interfaces that are going to just end up being halfway copies in naming and scheme of osu!lazer. A simple apply function works for us as-is, and there's no reason we can't use it.

osu!lazer has reason to, as it is split across reflected rulesets that have to import such interfaces and be reasonably passed over to the main game to easily differentiate what a mod does and doesnt do, while we can just apply the same real logic and basis to every mod.

Reply:
Before making comments like these, please consider the effort I take to ***avoid*** lazer.

Reply:
This isn't something to "copy" lazer, this is something purely out of convenience instead of having an exaggerated number of functions in a single class, it would look too jarring. Even if osu!lazer only had a single ruleset, I would imagine it'd still be based on what I just described.

You're taking the approach to avoid lazer's codebase too seriously, to the point where a design decision could be the death of an entire pull request, and while I can agree that we do share quite a few similarities in some areas of the codebase - or blatantly copied - design decisions such as this should not be qualified as "copying osu!lazer". That's plain and simply ridiculous.

Reply:
I come to the understanding that such seriousness must be upheld, as we would be violating the license of osu! by very virtue of similar design decisions. We are not in a position as effectively `indie developers` to be making moves that overstep even slight boundaries regarding open source licensing, as licensing is a territory of legal footing that could give a developer the rights to sue any infringing developers that violate the terms of the license provided for their code. I treat these matters as seriously as I do because I want to avoid any possible unintended consequences that come from these matters, as due to our knowledge and footing with the osu!lazer codebase, it is more than likely we will come across scenarios that will invariably lead to us having code and naming schemes that feel a bit too like lazer, which will come less off and transformative work and end up being just similar code but with changes to suit out purposes.

These same terms still fall under the MIT license that osu!lazer uses, under the basis of how copyleft works. Copyleft is a permission granted by the MIT license under the condition that we can freely use the `copyrighted property` of the lazer codebase, just so as long as we retain it's original licensing. If we overstep this very boundary, it could even go far as to having our work be taken down or sued (as unlikely as it is to happen.)

I am completely unwilling to change my mindset on this, because the concept and inclinations surrounding it are very severe. As I have said over DMs, I would rather discard the entire pull request than to take any steps that bring my work closer to such in any way.

Reply:
Every indie open-source project should be struck down and or sued with that viewpoint, and creating open-source projects with little to no line-by-line plagiarism is *COMPLETELY* impossible. My point still stands with the fact that design decisions such as this shouldn't impede our ability to progress with this project we've worked so hard on. Companies will use similar if not the same design decisions in their code, it does not matter in the grand scheme of things. 

We're talking about a couple of interfaces that would benefit us in figuring out where to apply them.

Is this really how you want things to be? To run through each and every mod, call a function in a designated area, and waste computing power because of that?

Reply:
If you have the evidence to prove this applies to *nearly every* instance, then you'll have successfully largely proven my point wrong. However, it's more than likely we'll come across these problems as a project using *the exact same framework*, handling *the exact same logic* being built for a game in *the exact same genre* as osu!. It is not unfair of me to have a nigh-extreme ideology regarding this topic with that in mind. I do believe inevitably at some point, most forms of code will copy off of others in some way; but this issue is just plainly unavoidable with current circumstances not being in our favour.

Also, is any computing power really wasted if you can only select a few mods, then those functions get applied just about the same way as they do in lazer minus requiring interfaces to designate such? We can still circumstantially handle basic logic with the function itself, varying by the very mod as intended. All I'm doing is simply just taking such and making them into a single function on a mod that can be called when necessary.

Reply:
On a side note, they already do for that matter. This is a complete PR; the "overlay" is only there just to cache selected mods.

Reply:
Fine.